trainer/finetune.py

```diff
@@ -0,0 +1,240 @@
+
+
+def main(argv):
+  """Main method for fintuning: builds, trains, and evaluates t5."""
+  tf.disable_v2_behavior()
+  tf.compat.v1.enable_eager_execution()
+
+  global PRETRAINED_DIR
+  global MODEL_DIR
+  PRETRAINED_DIR = os.path.join(BASE_PRETRAINED_DIR, FLAGS.size)
+  MODEL_DIR = os.path.join(MODELS_DIR, FLAGS.size, FLAGS.name)
+
+  print("--------DETECTING TPUs----")
+  try:
+    TPU_TOPOLOGY = "2x2"
+    tpu = tf.distribute.cluster_resolver.TPUClusterResolver()  # TPU detection
+    TPU_ADDRESS = tpu.get_master()
+    print('Running on TPU:', TPU_ADDRESS)
+  except ValueError:
+    raise BaseException('ERROR: Not connected to a TPU runtime')
+
+  # load or build data
+  if tf.io.gfile.exists(rd_counts_path):
+    print("TSV's Found")
+    # Used cached data and counts.
+    tf.logging.info("Loading Redial from cache.")
+    num_rd_examples = json.load(tf.io.gfile.GFile(rd_counts_path))
+  else:
+    print("TSV's Not Found")
+    # Create TSVs and get counts.
+    tf.logging.info("Generating Redial TSVs.")
+    num_rd_examples = {}
+    for split, fname in RD_SPLIT_FNAMES.items():
+      print(os.path.join(RD_JSONL_DIR, fname))
+      num_rd_examples[split] = preprocessing.rd_jsonl_to_tsv(
+          os.path.join(RD_JSONL_DIR, fname), rd_tsv_path[split])
+    json.dump(num_rd_examples, tf.io.gfile.GFile(rd_counts_path, "w"))
+
+  t5.data.TaskRegistry.add(
+      "rd_recommendations",
+      # Supply a function which returns a tf.data.Dataset.
+      dataset_fn=preprocessing.rd_dataset_fn,
+      splits=["train", "validation"],
+      # Supply a function which preprocesses text from the tf.data.Dataset.
+      text_preprocessor=[preprocessing.conversation_preprocessor],
+      # Use the same vocabulary that we used for pre-training.
+      sentencepiece_model_path=t5.data.DEFAULT_SPM_PATH,
+      # Lowercase targets before computing metrics.
+      postprocess_fn=t5.data.postprocessors.lower_text, 
+      # We'll use accuracy as our evaluation metric.
+      metric_fns=[t5.evaluation.metrics.accuracy, t5.evaluation.metrics.bleu],
+      # Not required, but helps for mixing and auto-caching.
+      num_input_examples=num_rd_examples)
+
+  # Set parallelism and batch size to fit on v2-8 TPU (if possible).
+  # Limit number of checkpoints to fit within 5GB (if possible).
+  model_parallelism, train_batch_size, keep_checkpoint_max = {
+      "small": (1, 256, 16),
+      "base": (2, 128, 8),
+      "large": (8, 64, 4),
+      "3B": (8, 16, 1),
+      "11B": (8, 16, 1)}[FLAGS.size]
+
+  tf.io.gfile.makedirs(MODEL_DIR)
+  # The models from the t5 paper paper are based on the Mesh Tensorflow Transformer.
+  model = t5.models.MtfModel(
+      model_dir=MODEL_DIR,
+      tpu=TPU_ADDRESS,
+      tpu_topology=TPU_TOPOLOGY,
+      model_parallelism=model_parallelism,
+      batch_size=train_batch_size,
+      sequence_length={"inputs": INPUT_LENGTH, "targets": TARGET_LENGTH},
+      learning_rate_schedule=0.003,
+      save_checkpoints_steps=5000,
```

Comment:
save_checkpoints and iterations_per_loop can potentially be dropped - these seem to be the default values anyway?

Reply:
I decided to change the save checkpoints to 2000 instead of 5000 because it seems like we're going to be training over 5-15k steps and this way we'll have more data on how the evaluation metrics are progressing over time. I got rid of iterations per loop.